pkg/config/template.go

```diff
@@ -47,3 +64,74 @@ func getTemplates(templatePath string) (CiTemplates, error) {
 func isYAML(file string, info os.FileInfo) bool {
 	return !info.IsDir() && (filepath.Ext(file) == ".yaml" || filepath.Ext(file) == ".yml")
 }
+
+// TemplateCMController holds the details needed for the configmap controller
+type TemplateCMController struct {
+	cmclient  corev1.ConfigMapInterface
+	prNumber  int
+	logger    *logrus.Entry
+	templates CiTemplates
+}
+
+// NewTemplateCMController creates a new TemplateCMController
+func NewTemplateCMController(cmclient corev1.ConfigMapInterface, prNumber int, logger *logrus.Entry, templates CiTemplates) *TemplateCMController {
+	return &TemplateCMController{
+		cmclient:  cmclient,
+		prNumber:  prNumber,
+		logger:    logger,
+		templates: templates,
+	}
+}
+
+// CreateCMTemplates creates configMaps for all the changed templates.
+func (c *TemplateCMController) CreateCMTemplates() error {
+	var errors []error
+	s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme)
+
+	for filename, template := range c.templates {
+		buf := new(bytes.Buffer)
+
+		err := s.Encode(template, buf)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		cmName := GetTempCMName(c.prNumber, template.Name)
+		cm := &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: cmName,
+				Labels: map[string]string{
+					rehearseLabel: fmt.Sprintf("rehearse-temporary-cm-%d", c.prNumber),
+				},
+			},
+			Data: map[string]string{filename: buf.String()},
+		}
+
+		c.logger.WithFields(logrus.Fields{"template-name": template.Name, "cm-name": cmName}).Info("creating rehearsal configMap for template")
+		if _, err := c.cmclient.Create(cm); err != nil {
+			if kerrors.IsAlreadyExists(err) {
+				if _, err := c.cmclient.Update(cm); err != nil {
+					errors = append(errors, fmt.Errorf("could not update existing configmap: %v", err))
+				}
+			} else {
+				errors = append(errors, err)
+			}
+		}
+	}
+	return kutilerrors.NewAggregate(errors)
+}
+
+// CleanupCMTemplates deletes all the configMaps that have been created for the changed templates.
+func (c *TemplateCMController) CleanupCMTemplates() error {
+	c.logger.Info("deleting temporary template configMaps")
+	if err := c.cmclient.DeleteCollection(&metav1.DeleteOptions{},
+		metav1.ListOptions{LabelSelector: fields.Set{rehearseLabel: fmt.Sprintf("rehearse-temporary-cm-%d", c.prNumber)}.AsSelector().String()}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetTempCMName converts a template name to a temporary name including the pr number.
+func GetTempCMName(prNumber int, templateName string) string {
```

Comment:
if we hash the data that is used to fill in the configmap, we will ensure we never have collisions and cleanup can be best-effort or even out-of-band, that seems more robust than this approach

Reply:
Hashing the data seems the best solution since recently I discovered that if someone run the tool with `--local` it will always create a cm with the same name.

Reply:
If you hash and collide, don't bother with anything -- a collision on the hash of the contents means that the CM in the cluster already has the right data in it

Reply:
that it a good approach. I will make the changes in this PR